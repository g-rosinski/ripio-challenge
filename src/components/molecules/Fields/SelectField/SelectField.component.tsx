import React, { InputHTMLAttributes } from "react"
import useInputHelper from "../../../../hooks/useInputHelper/useInputHelper.hook"
import { InputGroup, Label, HelperText } from "../../../atoms"
import { Select } from "./SelectField.styles"

type SelectFieldProps = InputHTMLAttributes<HTMLSelectElement> & {
    options: string[],
    // onChange: (value:string) => void,
    label?: string,
    helper?: string,
    error?: string,
}

const SelectField: React.FC<SelectFieldProps> = ({
    options, 
    label, 
    helper = "", 
    error,
    disabled,
    name,
    ...props
}) => {
    const { displayHelperText, helperText, helperStatus } = useInputHelper({
        disabled,
        error,
        info: helper
    })

    return (
        <InputGroup>
            { !!label && <Label htmlFor={name}>{label}</Label> }
            <Select {...props} id={name} disabled={disabled}>
                { 
                    options.map((option, i) => 
                        <option key={i} value={option}>{option}</option>
                    )
                }
            </Select>
            { displayHelperText && <HelperText status={helperStatus}>{ helperText }</HelperText> }
        </InputGroup>
    )
}

export default SelectField