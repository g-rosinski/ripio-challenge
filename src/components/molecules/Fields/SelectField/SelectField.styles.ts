import styled from "styled-components";

export const Select = styled.select`
    font-size: 14px;
    padding: 3px 5px;
    background: none;
    border: none;
    border-bottom: 1px solid;
    min-width: 100px;
`