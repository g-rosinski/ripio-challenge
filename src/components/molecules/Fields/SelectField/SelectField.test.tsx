import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import SelectField from './SelectField.component'

describe('<SelectField/>', () => {

    const onChangeMock = jest.fn()
    const optionsMock = ['DAI','USDT','BTH','ETH'].map(c => ({value: c}))

    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('Se espera que se reenderice correctamente', () => {
        
       const { container } = render(<SelectField options={optionsMock} onChange={onChangeMock} />)
        
        expect(screen.getByRole('combobox')).toHaveValue(optionsMock[0].value)
        expect(screen.queryAllByRole('option').length).toBe(4)
        expect(container).toMatchSnapshot()
    })

    test('Se espera que prop onChange al seleccionar otro valor', async () => {
        render(<SelectField options={optionsMock} onChange={onChangeMock} />)
        
        const select = screen.getByRole('combobox')
        
        userEvent.selectOptions(select,['USDT'])
        expect(onChangeMock).toHaveBeenCalledTimes(1)
        expect(select).toHaveValue('USDT')
    })

    
    test('Se espera que si disabled es true, la prop onChange no sea llamada', async () => {
        render(<SelectField options={optionsMock} onChange={onChangeMock} disabled />)
        
        const select = screen.getByRole('combobox')
        
        userEvent.selectOptions(select,['USDT'])
        expect(onChangeMock).not.toBeCalled()
        expect(select).toHaveValue(optionsMock[0].value)
    })

    test('Se espera que prop value se cambie al setear un value nuevo', async () => {
        const {rerender} = render(<SelectField options={optionsMock} value={""} onChange={onChangeMock} />)
        
        const select = screen.getByRole('combobox')
        
        expect(select).toHaveValue(optionsMock[0].value)

        optionsMock.forEach(option => {
            rerender(<SelectField options={optionsMock} value={option.value} onChange={onChangeMock} />)
            expect(select).toHaveValue(option.value)
        })
    })

    test('Se espera que al setear un label se reenderice un label', async () => {
        render(<SelectField options={optionsMock} label='Moneda' onChange={onChangeMock} />)
        
        expect(screen.getByText('Moneda')).toBeInTheDocument()
    })

    test('Se espera que al setear un helper se reenderice el helperText', async () => {
        render(<SelectField options={optionsMock} helper='Seleccione una moenda' onChange={onChangeMock} />)
        
        expect(screen.getByText('Seleccione una moenda')).toBeInTheDocument()
    })

    test('Se espera que al setear un error, helper se reenderice el error en el helperText', async () => {
        render(<SelectField options={optionsMock} error='Nombre no valido' helper='Ingrese un nombre' onChange={onChangeMock} />)
        
        expect(screen.getByText('Nombre no valido')).toBeInTheDocument()
        expect(screen.queryByText('Ingrese un nombre')).not.toBeInTheDocument()
    })

})