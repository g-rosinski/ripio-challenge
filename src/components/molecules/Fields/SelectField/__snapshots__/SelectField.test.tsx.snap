// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectField/> Se espera que se reenderice correctamente 1`] = `
<div>
  <div
    class="sc-gswNZR fGuKlY"
  >
    <select
      class="sc-gKPRtg jmYarQ"
    >
      <option
        value="DAI"
      >
        DAI
      </option>
      <option
        value="USDT"
      >
        USDT
      </option>
      <option
        value="BTH"
      >
        BTH
      </option>
      <option
        value="ETH"
      >
        ETH
      </option>
    </select>
  </div>
</div>
`;
